lib/Dialect/LLHD/IR/LLHDDialect.cpp

```diff
@@ -0,0 +1,374 @@
+  using DialectInlinerInterface::DialectInlinerInterface;
+
+  //===--------------------------------------------------------------------===//
+  // Analysis Hooks
+  //===--------------------------------------------------------------------===//
+
+  /// All operations within LLHD can be inlined.
+  bool isLegalToInline(Operation *, Region *,
+                       BlockAndValueMapping &) const final {
+    return true;
+  }
+
+  bool isLegalToInline(Region *, Region *src,
+                       BlockAndValueMapping &) const final {
+    // Don't inline processes and entities
+    return !isa<llhd::ProcOp>(src->getParentOp()) &&
+           !isa<llhd::EntityOp>(src->getParentOp());
+  }
+};
+} // end anonymous namespace
+
+//===----------------------------------------------------------------------===//
+// LLHD Dialect
+//===----------------------------------------------------------------------===//
+
+LLHDDialect::LLHDDialect(mlir::MLIRContext *context)
+    : Dialect(getDialectNamespace(), context) {
+  addTypes<SigType, TimeType>();
+  addAttributes<TimeAttr>();
+  addOperations<
+#define GET_OP_LIST
+#include "circt/Dialect/LLHD/IR/LLHD.cpp.inc"
+      >();
+  addInterfaces<LLHDInlinerInterface>();
+}
+
+Operation *LLHDDialect::materializeConstant(OpBuilder &builder, Attribute value,
+                                            Type type, Location loc) {
+  return builder.create<llhd::ConstOp>(loc, type, value);
+}
+
+//===----------------------------------------------------------------------===//
+// Type parsing
+//===----------------------------------------------------------------------===//
+
+/// Parse a signal type.
+/// Syntax: sig ::= !llhd.sig<type>
+static Type parseSigType(DialectAsmParser &parser) {
+  Type underlyingType;
+  if (parser.parseLess())
+    return Type();
+
+  llvm::SMLoc loc = parser.getCurrentLocation();
+  if (parser.parseType(underlyingType)) {
+    parser.emitError(loc, "No signal type found. Signal needs an underlying "
+                          "type.");
+    return nullptr;
+  }
+
+  if (parser.parseGreater())
+    return Type();
+  return SigType::get(underlyingType);
+}
+
+Type LLHDDialect::parseType(DialectAsmParser &parser) const {
+  llvm::StringRef typeKeyword;
+  // parse the type keyword first
+  if (parser.parseKeyword(&typeKeyword))
+    return Type();
+  if (typeKeyword == SigType::getKeyword()) {
+    return parseSigType(parser);
+  }
+  if (typeKeyword == TimeType::getKeyword())
+    return TimeType::get(getContext());
+  return Type();
+}
+
+//===----------------------------------------------------------------------===//
+// Type printing
+//===----------------------------------------------------------------------===//
+
+/// Print a signal type with custom syntax:
+/// type ::= !sig.type<underlying-type>
+static void printSigType(SigType sig, DialectAsmPrinter &printer) {
+  printer << sig.getKeyword() << "<";
+  printer.printType(sig.getUnderlyingType());
```

Comment:
nit: You can stream the type directly.

printer << ... << sig.getUnderlyingType() << ">";